File: Expense_BE/src/main/java/com/hunglevi/expense_be/repository/CategoryRepository.java
package com.hunglevi.expense_be.repository;

import com.hunglevi.expense_be.model.Category;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface CategoryRepository extends JpaRepository<Category, Integer> {

    @Query("SELECT c FROM categories c ORDER BY c.name ASC")
    List<Category> getAllCategories();

    @Query("SELECT c FROM categories c WHERE c.user.id = :userId ORDER BY c.name ASC")
    List<Category> getAllCategoriesByUserId(@Param("userId") Integer userId);

    @Query("SELECT c FROM categories c WHERE c.id = :id")
    Category getCategoryById(@Param("id") Integer id);

    @Query("SELECT c FROM categories c WHERE c.name LIKE %:searchTerm%")
    List<Category> searchCategoriesByName(@Param("searchTerm") String searchTerm);
}

File: Expense_MDC/app/src/main/java/com/hunglevi/expense_mdc/data/repository/CategoryRepository.kt
package com.hunglevi.expense_mdc.data.repository

import com.hunglevi.expense_mdc.data.dao.CategoryDao
import com.hunglevi.expense_mdc.data.model.Category
import kotlinx.coroutines.flow.Flow

class CategoryRepository(private val categoryDao: CategoryDao) {
    val allCategories: Flow<List<Category>> = categoryDao.getAllCategories()

    fun getCategoriesByUserId(userId: Int): Flow<List<Category>> {
        return categoryDao.getAllCategoriesByUserId(userId)
    }
    suspend fun insertCategory(category: Category) {
        categoryDao.insertCategory(category)
    }

    suspend fun deleteCategory(category: Category) {
        categoryDao.deleteCategory(category)
    }

    suspend fun getCategoryById(id: Int): Category? {
        return categoryDao.getCategoryById(id)
    }
}